### DISCUSSION TITLE
Migration from level 2.1.0 to 3.3.3

### DISCUSSION THREAD

**Discussion by RMerle** [forntoh/LcdMenu#141] [General]
I have been using LcdMenu Level 2.1.0 for 2 years in the PlatformIO environment and was very pleased with it. Unfortunately, I have updated it to level 3.3.3. I got so many errors on my code during the Build phase of PlatformIO that I decided to restart from scratch by building a new application using the Exemples supplied. I  got  many problems not with the exemples themselves but with the src MenuItem.h which fall into error for various reasons during the Build platformIO phase. I suspect this .h to be incompatible with platformIO. 
As I do not feel authorized to modify a .h myself,  I oppenned an Issue and appologize if this was not the right approach to tackle the problem.
However, I beleive that :
- To modify LcdMenu to this extent without giving any clue to the users of old version to make the move is questionnable.
- Not making sure that the new levels were running on PlatformIO and  leaving users of this Platform without any advice is questionnable as well.
Not getting any help will force me to stay with the old level or switch to an other tool.

**Comment by forntoh** [comment]
@RMerle Thanks for the remarks, this library is my hobby project, I do it because I love it.
So, I only work on improvements when I have enough time and have done the other jobs that pay me in order to have time for my own hobbies.

There have been multiple updates since v2.1.0 and if you progressively updated your, if you subscribed to the repo or starred it, you would have seen all the updates when they happen, I don't know all the users and it's up to them to keep their code up-to-date just the same way as I regularly update the library, I didn't just jump from v2.1.0 to v3.2.3, there were multiple versions in-between.

I have fully tested it with PlatformIO and I use PlatformIO to build this library and I didn't face any issues till now. Multiple other users use it with no major issues, of course, there might be unknown issues but I don't have the time to investigate every issue, that's why this project is open source, if you feel capable, you can investigate and submit a PR, I'll gladly review it.

Also, try to update your code progressively, I usually mention all changes in the release notes of every version.
You'll also be able to see how the repo looked like in every release, use that to see how the examples have been changing since 2.1.0

**Comment by RMerle** [comment]
Sir, I agree to almost  all what you say. I did not enough pay attention to the fact that LcdMenu belongs to the open world, a world that I am not familiar with (I am a retired electronic enginneer much more knowledgeable in the hardware part of the IS business). The reason why I did not touch  LCDMenu during two years is that the project in which I had incorporated  it  was not used during that time and has only be reactivated recently. However, when I realized the extent of what has been changed I tried to find some info on the various releases and could not find any except for the most reecent ones which was helpless.
As regard your proposal to  investigate on a solution, I would feel confortable if the issue concerned the Exemples code, which is not the case. What is at at stake is a component of the LCS Menu itself which falls into error as soon as PlatformIO invoques it. To undetake such a task, you need to know the architecture of the whole programm and the interaction of its various elements.
If you decide to leave the situation like it is, soon or later other users will fall into the same bug in the PIO environment and your user base will decline which is a pity in view of the quality of the job done and the effort you have already made.
